Signal/src/ViewControllers/ConversationView/ConversationViewModel.m

```diff
@@ -0,0 +1,1284 @@
+        [transformedItemList insertObject:itemId atIndex:newIndex];
+    }
+
+    if (![newItemIdList isEqualToArray:transformedItemList]) {
+        // We should be able to represent all transformations as a series of
+        // inserts, updates and deletes - moves should not be necessary.
+        //
+        // TODO: The unread indicator might end up being an exception.
+        OWSLogWarn(@"New and updated view item lists don't match.");
+        return [self.delegate conversationViewModelDidUpdate:ConversationUpdate.reloadUpdate];
+    }
+
+    // 3. Updates.
+    //
+    // NOTE: Order doesn't matter.
+    for (NSString *itemId in updatedItemSet) {
+        if (![newItemIdList containsObject:itemId]) {
+            OWSFailDebug(@"Updated view item not in new view item list.");
+            continue;
+        }
+        if ([insertedItemIdList containsObject:itemId]) {
+            continue;
+        }
+        NSUInteger oldIndex = [oldItemIdList indexOfObject:itemId];
+        if (oldIndex == NSNotFound) {
+            OWSFailDebug(@"Can't find old index of updated view item.");
+            return [self.delegate conversationViewModelDidUpdate:ConversationUpdate.reloadUpdate];
+        }
+        NSUInteger newIndex = [newItemIdList indexOfObject:itemId];
+        if (newIndex == NSNotFound) {
+            OWSFailDebug(@"Can't find new index of updated view item.");
+            return [self.delegate conversationViewModelDidUpdate:ConversationUpdate.reloadUpdate];
+        }
+        id<ConversationViewItem> _Nullable viewItem = newViewItemMap[itemId];
+        if (!viewItem) {
+            OWSFailDebug(@"Can't find inserted view item.");
+            return [self.delegate conversationViewModelDidUpdate:ConversationUpdate.reloadUpdate];
+        }
+        [updateItems addObject:[[ConversationUpdateItem alloc] initWithUpdateItemType:ConversationUpdateItemType_Update
+                                                                             oldIndex:oldIndex
+                                                                             newIndex:newIndex
+                                                                             viewItem:viewItem]];
+    }
+
+    BOOL shouldAnimateUpdates = [self shouldAnimateUpdateItems:updateItems
+                                              oldViewItemCount:oldViewItemCount
+                                        updatedNeighborItemSet:updatedNeighborItemSet];
+
+    return [self.delegate
+        conversationViewModelDidUpdate:[ConversationUpdate diffUpdateWithUpdateItems:updateItems
+                                                                shouldAnimateUpdates:shouldAnimateUpdates]];
+}
+
+- (BOOL)shouldAnimateUpdateItems:(NSArray<ConversationUpdateItem *> *)updateItems
+                oldViewItemCount:(NSUInteger)oldViewItemCount
+          updatedNeighborItemSet:(nullable NSMutableSet<NSString *> *)updatedNeighborItemSet
+{
```

Comment:
I moved this here because the CVM had the necessary state to make this decision and I was reluctant to plumb it through the "CVM update" event.  However, animations are a view concern and probably don't belong in the CVM.